deploy/ciao-deploy-client/main.go

```diff
@@ -0,0 +1,474 @@
+	flag.StringVar(&token, "token", "ciao", "Secret token shared with server")
+	flag.StringVar(&role, "role", "", "One of: controller, agent, netagent or dual (agent and netagent)")
+	flag.BoolVar(&downloadClient, "download-client", true, "Download client from server")
+	flag.BoolVar(&force, "force", false, "Force steps that would normally skippped")
+	flag.StringVar(&systemPkiLocation, "system-pki-location", "/etc/pki/ciao", "Filesystem location for certificates")
+}
+
+func roleToSSNTPRole(role string) ssntp.Role {
+	switch role {
+	case "agent":
+		return ssntp.AGENT
+	case "netagent":
+		return ssntp.NETAGENT
+	case "dual":
+		return ssntp.NETAGENT | ssntp.AGENT
+	case "controller":
+		return ssntp.Controller
+	default:
+		return ssntp.UNKNOWN
+	}
+}
+
+func validateFlags() error {
+	if deployServerHost == "" {
+		return errors.New("--deploy-server must be specified")
+	}
+
+	switch role {
+	case "agent":
+	case "netagent":
+	case "controller":
+	case "dual":
+	default:
+		return errors.New("--role must be one of controller, agent, netagent or dual")
+	}
+
+	return nil
+}
+
+func deploymentDirectory() string {
+	u, err := user.Current()
+	d := "."
+	if err == nil {
+		d = u.HomeDir
+	}
+	return path.Join(d, ".ciao-deployment")
+}
+
+func csrPath(ssntpRole ssntp.Role) string {
+	return path.Join(deploymentDirectory(), deploy.CSRName(ssntpRole))
+}
+
+func privPath(ssntpRole ssntp.Role) string {
+	return path.Join(deploymentDirectory(), deploy.PrivName(ssntpRole))
+}
+
+// Generate and upload. Saving to disk for subsequent certificate fetching.
+func uploadCSR() (fp string, err error) {
+	var csrOutput, privKeyOutput bytes.Buffer
+	hs, err := os.Hostname()
+	if err != nil {
+		hs = "localhost"
+	}
+
+	hosts := []string{hs}
+
+	request := certs.CreateCertificateRequest(false, "Ciao Deployment", "", hosts, nil)
+
+	err = certs.CreateCSR(request, false, &csrOutput, &privKeyOutput)
+	if err != nil {
+		return "", errors.Wrap(err, "Error creating CSR")
+	}
+
+	csrBlock, _ := pem.Decode(csrOutput.Bytes())
+	certReq, _ := x509.ParseCertificateRequest(csrBlock.Bytes)
+	fp = certs.FingerPrint(certReq)
+	fmt.Printf("Requesting CSR with fingerprint: %s\n", fp)
+
+	url := fmt.Sprintf("%s/sign/%s/%s", deployServerHost, token, role)
+	fmt.Printf("Uploading CSR to %s\n", url)
+
+	defer func() {
+		if err != nil {
+			os.Remove(csrPath(ssntpRole))
+			os.Remove(privPath(ssntpRole))
+		}
+	}()
+
+	err = ioutil.WriteFile(csrPath(ssntpRole), csrOutput.Bytes(), 0600)
+	if err != nil {
+		return "", errors.Wrap(err, "Error saving CSR to file")
+	}
+
+	err = ioutil.WriteFile(privPath(ssntpRole), privKeyOutput.Bytes(), 0600)
+	if err != nil {
+		return "", errors.Wrap(err, "Error saving private key to file")
+	}
+
+	r, err := http.NewRequest("PUT", url, &csrOutput)
+	if err != nil {
+		return "", errors.Wrap(err, "Error requesting signing of CSR")
+	}
+
+	r.Header.Set("Content-Type", "text/plain")
+
+	client := &http.Client{}
+	res, err := client.Do(r)
```

Comment:
As a general comment there are a few long running networky tasks, and invocations of other tasks, launched from this tool, with no way to cancel them.  If one of these hangs will this prevent me from being able to cleanly quit ./setup.sh?